File: src/main/scala/com/dindane/mireille/visitors/InvokeDynamicVisitor.java
package main.scala.com.dindane.mireille.visitors;

import org.objectweb.asm.*;
import main.scala.com.dindane.mireille.models.InvokeDynamicCall;
import scala.Option;
import scala.Some;
import scala.collection.Seq;
import java.util.ArrayList;
import java.util.Arrays;

public class InvokeDynamicVisitor extends ClassVisitor {
    public ArrayList<InvokeDynamicCall> invokeDynamicCalls = new ArrayList<InvokeDynamicCall>();
    private String className;
    private Option<String> fileName;

    public InvokeDynamicVisitor(String className) {
        super(Opcodes.ASM4);
        this.className = className;
    }

    @Override
    public void visitSource(String source, String debug) {
        fileName = new Some(source);
    }

    public MethodVisitor visitMethod(int access, String name, final String description, String signature, String[] exceptions) {
        return new MethodVisitor(Opcodes.ASM4) {
            int lineNumber;

            public void visitLineNumber(int line, Label startLabel) {
                lineNumber = line;
            }

            public void visitInvokeDynamicInsn(String name, String desc, Handle bsm, Object... bsmArgs) {
                Seq<Type> arguments = scala.collection.JavaConversions.asScalaBuffer(
                        Arrays.asList(Type.getArgumentTypes(description))
                ).seq();

                invokeDynamicCalls.add(new InvokeDynamicCall(className,
                        name,
                        arguments,
                        Type.getReturnType(description),
                        fileName,
                        new Some(lineNumber)));
            }
        };
    }
}

File: src/main/scala/com/dindane/mireille/visitors/InvokeDynamicVisitor.scala
package main.scala.com.dindane.mireille.visitors

import collection._
import org.objectweb.asm._
import main.scala.com.dindane.mireille.models.InvokeDynamicCall

/**
 * A class visitor that visits a Java class and stores its Invoke Dynamic calls
 * in the <code>invokeDynamicCalls</code> List
 *
 * @param className The class to be visited
 */
class InvokeDynamicVisitor(className: String) extends ClassVisitor(Opcodes.ASM4) {

  val invokeDynamicCalls: mutable.MutableList[InvokeDynamicCall] = mutable.MutableList.empty
  private var fileName: Option[String] = None

  override def visitSource(source: String, debug: String) {
    fileName = Some(source)
  }

  override def visitMethod(access: Int, name: String, description: String, signature: String, exceptions: Array[String]): MethodVisitor = {
    new MethodVisitor(Opcodes.ASM4) {
      private var lineNumber: Option[Int] = None

      override def visitLineNumber(line: Int, startLabel: Label) {
        lineNumber = Some(line)
      }

      override def visitInvokeDynamicInsn(name: String, desc: String, bsm: Handle, bsmArgs: Object*) {
        invokeDynamicCalls += InvokeDynamicCall(className,
          name,
          Type.getArgumentTypes(description),
          Type.getReturnType(description),
          fileName,
          lineNumber,
          bsm)
      }
    }
  }

}
